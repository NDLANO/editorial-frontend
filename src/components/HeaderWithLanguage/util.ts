/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { Node } from "@ndla/types-taxonomy";
import {
  toEditAudio,
  toEditConcept,
  toEditFrontPageArticle,
  toEditGloss,
  toEditImage,
  toEditLearningResource,
  toEditPodcast,
  toEditPodcastSeries,
  toEditTopicArticle,
} from "../../util/routeHelpers";

export const getTaxonomyPathsFromTaxonomy = (taxonomy?: Pick<Node, "paths">[], articleId?: number): string[] => {
  const taxPaths = taxonomy?.flatMap((t) => t.paths) ?? [];
  const articlePath = articleId ? `/article/${articleId}` : undefined;
  const paths = articlePath ? taxPaths.concat(articlePath) : taxPaths;
  return paths.filter((p): p is string => p !== undefined);
};

export const toMapping = {
  concept: toEditConcept,
  gloss: toEditGloss,
  audio: toEditAudio,
  "podcast-series": toEditPodcastSeries,
  podcast: toEditPodcast,
  image: toEditImage,
  "frontpage-article": toEditFrontPageArticle,
  standard: toEditLearningResource,
  "topic-article": toEditTopicArticle,
};

export type TranslatableType = keyof typeof toMapping;

export const translatableTypes: TranslatableType[] = [
  "audio",
  "concept",
  "gloss",
  "standard",
  "topic-article",
  "podcast",
  "image",
  "podcast-series",
  "frontpage-article",
];

export const createEditUrl = (id: number, locale: string, type: keyof typeof toMapping) => {
  return toMapping[type](id, locale);
};
