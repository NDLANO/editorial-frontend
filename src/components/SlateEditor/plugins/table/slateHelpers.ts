/**
 * Copyright (c) 2022-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *
 */

import { Editor, Element, Node } from 'slate';
import { jsx as slatejsx } from 'slate-hyperscript';
import { defaultTableCellBlock } from './defaultBlocks';
import {
  TableBodyElement,
  TableCaptionElement,
  TableCellElement,
  TableElement,
  TableHeadElement,
  TableRowElement,
} from './interfaces';
import {
  TYPE_TABLE,
  TYPE_TABLE_BODY,
  TYPE_TABLE_CAPTION,
  TYPE_TABLE_CELL,
  TYPE_TABLE_HEAD,
  TYPE_TABLE_ROW,
} from './types';

export const isTable = (node?: Node): node is TableElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE;
};

export const isTableCaption = (node?: Node): node is TableCaptionElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_CAPTION;
};
export const isTableHead = (node?: Node): node is TableHeadElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_HEAD;
};

export const isTableBody = (node?: Node): node is TableBodyElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_BODY;
};

export const isTableRow = (node?: Node): node is TableRowElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_ROW;
};

export const isTableCell = (node?: Node): node is TableCellElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_CELL;
};

export const hasCellAlignOfType = (editor: Editor, type: string) => {
  // For all selected table cells
  for (const [cell] of Editor.nodes<TableCellElement>(editor, {
    match: node => isTableCell(node),
  })) {
    if (cell.data.align === type) {
      return true;
    }
  }
  return false;
};

export const countCells = (row: TableRowElement, stop?: number) => {
  return row.children
    .map(child => {
      if (!isTableCell(child)) {
        return 0;
      }
      return child.data.colspan;
    })
    .slice(0, stop)
    .reduce((a, b) => a + b);
};

export const getTableBodyWidth = (element: TableHeadElement | TableBodyElement) => {
  const firstRow = element.children[0];
  if (isTableRow(firstRow)) {
    return countCells(firstRow);
  }
  return 0;
};

export const getTableBodyHeight = (element: TableHeadElement | TableBodyElement) => {
  return element.children.length;
};

export const createIdenticalRow = (element: TableRowElement) => {
  return slatejsx(
    'element',
    { type: TYPE_TABLE_ROW },
    element.children.map(child => {
      if (isTableCell(child)) {
        return {
          ...defaultTableCellBlock(),
          data: {
            ...child.data,
            rowspan: 1,
          },
        };
      }
      return defaultTableCellBlock();
    }),
  );
};
