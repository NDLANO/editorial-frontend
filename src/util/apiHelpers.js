/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/* eslint-disable no-shadow */

import defined from 'defined';
import fetch from 'isomorphic-fetch';
import queryString from 'query-string';
import config from '../config';
import {
  apiBaseUrl,
  getAccessToken,
  isAccessTokenValid,
  renewAuth,
} from './authHelpers';

export function apiResourceUrl(path) {
  return apiBaseUrl + path;
}

export function brightcoveApiResourceUrl(path) {
  return config.brightcoveApiUrl + path;
}

export function googleSearchApiResourceUrl(path) {
  return config.googleSearchApiUrl + path;
}

export function createErrorPayload(status, message, json) {
  return Object.assign(new Error(message), { status, json });
}

export function resolveJsonOrRejectWithError(res, taxonomy = false) {
  return new Promise((resolve, reject) => {
    if (res.ok) {
      if (res.status === 204) {
        return taxonomy ? resolve(true) : resolve();
      }
      // Temporary until API changes to return representation
      const location = res.headers.get('Location');
      if (res.status === 201 && (location || taxonomy)) {
        if (!location && taxonomy) {
          resolve();
        }
        return resolve(location);
      }
      return resolve(res.json());
    }
    return res
      .json()
      .then(json =>
        reject(
          createErrorPayload(
            res.status,
            defined(json.message, res.statusText),
            json,
          ),
        ),
      )
      .catch(reject);
  });
}

export const fetchWithAuthorization = async (url, config = {}, forceAuth) => {
  if (forceAuth || !isAccessTokenValid()) {
    await renewAuth();
  }
  return fetch(url, {
    ...config,
    headers: {
      Authorization: `Bearer ${getAccessToken()}`,
      'Content-Type': config.headers
        ? config.headers['Content-Type']
        : 'text/plain',
    },
  });
};

export const fetchAuthorized = (url, config = {}) =>
  fetchWithAuthorization(url, config, false);

export const fetchReAuthorized = async (url, config = {}) =>
  fetchWithAuthorization(url, config, true);

export const fetchBrightcoveAccessToken = () =>
  fetch('/get_brightcove_token').then(resolveJsonOrRejectWithError);

export const setBrightcoveAccessTokenInLocalStorage = brightcoveAccessToken => {
  localStorage.setItem(
    'brightcove_access_token',
    brightcoveAccessToken.access_token,
  );
  localStorage.setItem(
    'brightcove_access_token_expires_at',
    brightcoveAccessToken.expires_in * 1000 + new Date().getTime(),
  );
};

export const fetchWithBrightCoveToken = (url, config = {}) => {
  const brightcoveAccessToken = localStorage.getItem('brightcove_access_token');
  const expiresAt = brightcoveAccessToken
    ? JSON.parse(localStorage.getItem('brightcove_access_token_expires_at'))
    : 0;
  if (new Date().getTime() > expiresAt || !expiresAt) {
    return fetchBrightcoveAccessToken().then(res => {
      setBrightcoveAccessTokenInLocalStorage(res);
      return fetch(url, {
        ...config,
        headers: { Authorization: `Bearer ${res.access_token}` },
      });
    });
  }
  return fetch(url, {
    ...config,
    headers: { Authorization: `Bearer ${brightcoveAccessToken}` },
  });
};

export const fetchOembed = async (url, options) => {
  const data = await fetchAuthorized(url, options);
  return resolveJsonOrRejectWithError(data);
};

export const setOembedUrl = query =>
  `${apiResourceUrl('/oembed-proxy/v1/oembed')}?${queryString.stringify(
    query,
  )}`;

export const fetchExternalOembed = (url, options) =>
  fetchOembed(setOembedUrl({ url }), options);
