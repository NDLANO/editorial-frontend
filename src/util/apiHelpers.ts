/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import fetch from 'cross-fetch';
import queryString from 'query-string';
import { BrightcoveAccessToken, H5POembed } from '../interfaces';
import config from '../config';
import { apiBaseUrl, getAccessToken, isAccessTokenValid, renewAuth } from './authHelpers';
import { resolveJsonOrRejectWithError, throwErrorPayload } from './resolveJsonOrRejectWithError';

export interface HttpHeadersType {
  'Content-Type': string;
  Authorization: string;
  VersionHash: string;
}

export interface FetchConfigType {
  method?: 'POST' | 'PUT' | 'DELETE' | 'GET' | 'PATCH';
  headers?: Partial<HttpHeadersType>;
  body?: any;
}

export function apiResourceUrl(path: string) {
  return apiBaseUrl + path;
}

export function grepUrl(path = '') {
  return `${apiResourceUrl('/grep/kl06/v201906')}${path}`;
}

export function brightcoveApiResourceUrl(path: string) {
  return config.brightcoveApiUrl + path;
}

export function googleSearchApiResourceUrl(path: string) {
  return config.googleSearchApiUrl + path;
}

export const fetchWithAuthorization = async (
  url: string,
  config: FetchConfigType = {},
  forceAuth: boolean,
) => {
  if (forceAuth || !isAccessTokenValid()) {
    await renewAuth();
  }

  const contentType = config.headers ? config.headers['Content-Type'] : 'text/plain';
  const extraHeaders = contentType ? { 'Content-Type': contentType } : null;
  const cacheControl = { 'Cache-Control': 'no-cache' };
  const headers: HeadersInit = {
    ...extraHeaders,
    ...cacheControl,
    VersionHash: config.headers?.VersionHash ?? 'default',
    Authorization: `Bearer ${getAccessToken()}`,
  };

  return fetch(url, {
    ...config,
    headers,
  });
};

const defaultHeaders = { 'Content-Type': 'application/json' };

interface DoAndResolveType<Type> extends FetchConfigType {
  url: string;
  alternateResolve?: (res: Response) => Promise<Type>;
  taxonomyVersion?: string;
}

interface HttpConfig<T> extends Omit<DoAndResolveType<T>, 'method'> {}

interface FetchConfig<T> extends HttpConfig<T> {
  queryParams?: Record<string, any>;
}

const httpResolve = <Type>({
  url,
  headers,
  alternateResolve,
  taxonomyVersion,
  ...config
}: DoAndResolveType<Type>): Promise<Type> => {
  return fetchAuthorized(url, {
    ...config,
    headers: { ...defaultHeaders, VersionHash: taxonomyVersion, ...headers },
  }).then(r => {
    return alternateResolve?.(r) ?? resolveJsonOrRejectWithError(r);
  });
};

export const stringifyQuery = (object: Record<string, any> = {}) => {
  const stringified = `?${queryString.stringify(object)}`;
  return stringified === '?' ? '' : stringified;
};

export const httpFunctions = {
  postAndResolve: <T>(conf: HttpConfig<T>) => httpResolve<T>({ ...conf, method: 'POST' }),
  putAndResolve: <T>(conf: HttpConfig<T>) => httpResolve<T>({ ...conf, method: 'PUT' }),
  patchAndResolve: <T>(conf: HttpConfig<T>) => httpResolve<T>({ ...conf, method: 'PATCH' }),
  deleteAndResolve: <T>(conf: HttpConfig<T>) => httpResolve<T>({ ...conf, method: 'DELETE' }),
  fetchAndResolve: <T>(conf: FetchConfig<T>) =>
    httpResolve<T>({
      ...conf,
      method: 'GET',
      url: `${conf.url}${stringifyQuery(conf.queryParams)}`,
    }),
};

export const fetchAuthorized = (url: string, config: FetchConfigType = {}) =>
  fetchWithAuthorization(url, config, false);

export const fetchReAuthorized = async (url: string, config: FetchConfigType = {}) =>
  fetchWithAuthorization(url, config, true);

export const fetchBrightcoveAccessToken = () =>
  fetch('/get_brightcove_token').then(r => resolveJsonOrRejectWithError<BrightcoveAccessToken>(r));

export const setBrightcoveAccessTokenInLocalStorage = (
  brightcoveAccessToken: BrightcoveAccessToken,
) => {
  localStorage.setItem('brightcove_access_token', brightcoveAccessToken.access_token);
  localStorage.setItem(
    'brightcove_access_token_expires_at',
    (brightcoveAccessToken.expires_in * 1000 + new Date().getTime()).toString(),
  );
};

export const fetchWithBrightCoveToken = (url: string) => {
  const brightcoveAccessToken = localStorage.getItem('brightcove_access_token');
  const expiresAt = brightcoveAccessToken
    ? JSON.parse(localStorage.getItem('brightcove_access_token_expires_at')!)
    : 0;
  if (new Date().getTime() > expiresAt || !expiresAt) {
    return fetchBrightcoveAccessToken().then(res => {
      setBrightcoveAccessTokenInLocalStorage(res);
      return fetch(url, {
        headers: { Authorization: `Bearer ${res.access_token}` },
      });
    });
  }
  return fetch(url, {
    headers: { Authorization: `Bearer ${brightcoveAccessToken}` },
  });
};

export const fetchOembed = async (url: string, options?: FetchConfigType): Promise<H5POembed> => {
  const data = await fetchAuthorized(url, options);
  return resolveJsonOrRejectWithError(data);
};

const setH5pOembedUrl = (query: { url: string }) =>
  `${config.h5pApiUrl}/oembed/preview?${queryString.stringify(query)}`;

const setOembedUrl = (query: { url: string }) =>
  `${apiResourceUrl('/oembed-proxy/v1/oembed')}?${queryString.stringify(query)}`;

export const fetchExternalOembed = (url: string, options?: FetchConfigType) => {
  let setOembed = setOembedUrl({ url });
  if (config.h5pApiUrl && url.includes(config.h5pApiUrl)) {
    setOembed = setH5pOembedUrl({ url });
  }
  return fetchOembed(setOembed, options);
};

export { resolveJsonOrRejectWithError, throwErrorPayload };
