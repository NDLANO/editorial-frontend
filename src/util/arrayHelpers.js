/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

// swapping array elements
export function arrMove(array, fromIndex, toIndex) {
  const item = array[fromIndex];
  const length = array.length;
  const diff = fromIndex - toIndex;

  if (diff > 0) {
    return [
      ...array.slice(0, toIndex),
      item,
      ...array.slice(toIndex, fromIndex),
      ...array.slice(fromIndex + 1, length),
    ];
  } else if (diff < 0) {
    const targetIndex = toIndex + 1;
    return [
      ...array.slice(0, fromIndex),
      ...array.slice(fromIndex + 1, targetIndex),
      item,
      ...array.slice(targetIndex, length),
    ];
  }
  return array;
}
