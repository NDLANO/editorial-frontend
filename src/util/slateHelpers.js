/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import {
  reduceElementDataAttributes,
  createEmbedProps,
  reduceChildElements,
} from './embedTagHelpers';

export const BLOCK_TAGS = {
  section: 'section',
  blockquote: 'quote',
  details: 'details',
  summary: 'summary',
  pre: 'code',
  h1: 'heading-two',
  h2: 'heading-two',
  h3: 'heading-two',
  h4: 'heading-two',
  h5: 'heading-two',
  h6: 'heading-two',
  br: 'br',
};

export const TABLE_TAGS = {
  table: 'table',
  th: 'table-cell',
  tr: 'table-row',
  td: 'table-cell',
};

const MARK_TAGS = {
  strong: 'bold',
  em: 'italic',
  u: 'underlined',
  code: 'code',
};

const ListText = ({ children }) => children;

export const findNodesByType = (node, type, nodes = []) => {
  if (node.type === type) {
    nodes.push(node);
  } else if (
    node.kind === 'document' ||
    (node.kind === 'block' && node.nodes.size > 0)
  ) {
    node.nodes.forEach(n => findNodesByType(n, type, nodes));
  }
  return nodes;
};

export const toJSON = state => state.toJSON();

export const logState = state => {
  console.log(JSON.stringify(toJSON(state), null, 2)); // eslint-disable-line no-console
};

// TODO: get type of aside in here. Default should be rightAside since that is the only
const getAsideType = el => ({
  type: el.attributes.getNamedItem('data-type')
    ? el.attributes.getNamedItem('data-type').value
    : 'rightAside',
});

const setAsideTag = data => ({
  'data-type': data.get('type') || '',
});

const illegalTextUnderBlocks = ['ol', 'ul'];

/* eslint-disable consistent-return, default-case */
export const textRule = {
  deserialize(el) {
    if (
      el.nodeName.toLowerCase() === '#text' &&
      el.parentNode &&
      illegalTextUnderBlocks.includes(el.parentNode.tagName.toLowerCase())
    ) {
      return null;
    } else if (
      !el.nodeName ||
      el.nodeName.toLowerCase() !== '#text' ||
      (el.parentNode && el.parentNode.tagName.toLowerCase() !== 'section')
    ) {
      return;
    }
    return null;
  },
};
export const divRule = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'div') return;

    if (el.className === 'c-bodybox') {
      return {
        kind: 'block',
        type: 'bodybox',
        nodes: next(el.childNodes),
      };
    } else if (el.dataset.type === 'related-content') {
      return {
        kind: 'block',
        type: 'related',
        isVoid: true,
        data: reduceChildElements(el),
      };
    }
    const childs = next(el.childNodes);
    return {
      kind: 'block',
      type: 'div',
      nodes: childs,
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'div' && object.type !== 'bodybox') return;
    switch (object.type) {
      case 'bodybox':
        return <div className="c-bodybox">{children}</div>;
      default:
        return <div>{children}</div>;
    }
  },
};

export const paragraphRule = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'p') return;
    const parent = el.parentElement
      ? el.parentElement.tagName.toLowerCase()
      : '';

    const type = parent === 'li' ? 'list-text' : 'paragraph';

    return {
      kind: 'block',
      type,
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'paragraph' && object.type !== 'list-text') return;
    if (object.type === 'list-text') {
      return <ListText>{children}</ListText>;
    }
    return <p>{children}</p>;
  },
};

export const listItemRule = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'li') return;
    return {
      kind: 'block',
      type: 'list-item',
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'list-item') return;
    return <li>{children}</li>;
  },
};

export const unorderListRules = {
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'ul') return;
    const type = el.attributes.getNamedItem('data-type');
    const data = { type: type ? type.value : '' };

    if (data.type === 'two-column') {
      return {
        kind: 'block',
        type: 'two-column-list',
        nodes: next(el.childNodes),
        data,
      };
    }

    return {
      kind: 'block',
      type: 'bulleted-list',
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'two-column-list' && object.type !== 'bulleted-list') {
      return;
    }

    if (object.type === 'two-column-list') {
      return <ul data-type="two-column">{children}</ul>;
    }
    return <ul>{children}</ul>;
  },
};

export const orderListRules = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'ol') return;
    const type = el.attributes.getNamedItem('data-type');
    const data = { type: type ? type.value : '' };
    if (data.type === 'letters') {
      return {
        kind: 'block',
        type: 'letter-list',
        nodes: next(el.childNodes),
        data,
      };
    }
    return {
      kind: 'block',
      type: 'numbered-list',
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'numbered-list' && object.type !== 'letter-list')
      return;
    if (object.type === 'letter-list') {
      return <ol data-type="letters">{children}</ol>;
    }
    return <ol>{children}</ol>;
  },
};

export const footnoteRule = {
  deserialize(el) {
    if (!el.tagName.toLowerCase().startsWith('embed')) return;
    const embed = reduceElementDataAttributes(el);
    if (embed.resource !== 'footnote') return;

    return {
      kind: 'inline',
      type: 'footnote',
      nodes: [
        {
          kind: 'text',
          isVoid: true,
          leaves: [
            {
              kind: 'leaf',
              text: '#',
              marks: [],
            },
          ],
        },
      ],
      data: {
        ...embed,
        authors: embed.authors ? embed.authors.split(';') : [],
      },
    };
  },
  serialize(object) {
    if (object.kind !== 'inline') return;
    if (object.type !== 'footnote') return;

    const data = object.data.toJS();
    const props = createEmbedProps({
      ...data,
      authors: data.authors ? data.authors.join(';') : '',
    });
    return <embed {...props} />;
  },
};

export const blockRules = {
  deserialize(el, next) {
    const block = BLOCK_TAGS[el.tagName.toLowerCase()];
    if (!block) return;
    return {
      kind: 'block',
      type: block,
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    switch (object.type) {
      case 'section':
        return <section>{children}</section>;
      case 'bulleted-list':
        return <ul>{children}</ul>;
      case 'heading-one':
        return <h1>{children}</h1>;
      case 'heading-two':
        return <h2>{children}</h2>;
      case 'heading-three':
        return <h3>{children}</h3>;
      case 'heading-four':
        return <h4>{children}</h4>;
      case 'heading-five':
        return <h5>{children}</h5>;
      case 'heading-six':
        return <h6>{children}</h6>;
      case 'quote':
        return <blockquote>{children}</blockquote>;
      case 'details':
        return <details>{children}</details>;
      case 'summary':
        return <summary>{children}</summary>;
      case 'br':
        return <br />;
    }
  },
};

export const tableRules = {
  deserialize(el, next) {
    const tableTag = TABLE_TAGS[el.tagName.toLowerCase()];
    if (!tableTag) return;
    return {
      kind: 'block',
      type: tableTag,
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    switch (object.type) {
      case 'table': {
        return (
          <table>
            <thead>{children.slice(0, 1)}</thead>
            <tbody>{children.slice(1)}</tbody>
          </table>
        );
      }
      case 'table-row':
        return <tr>{children}</tr>;
      case 'table-cell':
        if (object.data.get('isHeader')) {
          return <th>{children}</th>;
        }
        return <td>{children}</td>;
    }
  },
};

const relatedRule = {
  serialize(object) {
    if (object.type === 'related') {
      return (
        <div data-type="related-content">
          {object.data.get('nodes') &&
            object.data
              .get('nodes')
              .map(node => <embed {...createEmbedProps(node)} />)}
        </div>
      );
    }
  },
};

const RULES = [
  divRule,
  textRule,
  orderListRules,
  unorderListRules,
  tableRules,
  paragraphRule,
  listItemRule,
  relatedRule,
  {
    // Aside handling
    deserialize(el, next) {
      if (el.tagName.toLowerCase() !== 'aside') return;
      return {
        kind: 'block',
        type: 'aside',
        nodes: next(el.childNodes),
        data: getAsideType(el),
      };
    },
    serialize(object, children) {
      if (object.kind !== 'block') return;
      if (object.type !== 'aside') return;
      return <aside {...setAsideTag(object.data)}>{children}</aside>;
    },
  },
  blockRules,
  {
    deserialize(el, next) {
      const mark = MARK_TAGS[el.tagName.toLowerCase()];
      if (!mark) return;
      return {
        kind: 'mark',
        type: mark,
        nodes: next(el.childNodes),
      };
    },
    serialize(object, children) {
      if (object.kind !== 'mark') return;
      switch (object.type) {
        case 'bold':
          return <strong>{children}</strong>;
        case 'italic':
          return <em>{children}</em>;
        case 'underlined':
          return <u>{children}</u>;
        case 'superscripted':
          return <sup>{children}</sup>;
      }
    },
  },
  footnoteRule,
  {
    deserialize(el, next) {
      if (el.tagName.toLowerCase() !== 'a') return;
      return {
        kind: 'inline',
        type: 'link',
        data: {
          href: el.href ? el.href : '#',
          target: el.target ? el.target : '',
          rel: el.rel ? el.rel : '',
        },
        nodes: next(el.childNodes),
      };
    },
    serialize(object, children) {
      if (object.kind !== 'inline') return;
      if (object.type !== 'link') return;
      const data = object.data.toJS();

      if (data.resource === 'content-link') {
        return (
          <embed
            data-resource={data.resource}
            data-content-id={data['content-id']}
            data-link-text={object.text}
            data-open-in={data['open-in']}
          />
        );
      }

      return (
        <a
          href={data.href}
          target={data.target}
          rel={data.rel}
          title={object.text}>
          {children}
        </a>
      );
    },
  },
];

const topicArticeEmbedRule = [
  {
    // Embeds handling
    deserialize(el) {
      if (el.tagName.toLowerCase() !== 'embed') return;

      if (el.dateset['data-resource'] === 'related-content') {
        return;
      }
      return {
        kind: 'block',
        type: 'embed',
        data: reduceElementDataAttributes(el),
        isVoid: true,
      };
    },
    serialize(object) {
      if (object.kind !== 'block') return;
      if (object.type !== 'embed') return;
      switch (object.type) {
        case 'embed':
          return <deleteme />;
      }
    },
  },
];

export const learningResourceEmbedRule = [
  {
    deserialize(el) {
      if (!el.tagName.toLowerCase().startsWith('embed')) return;

      const embed = reduceElementDataAttributes(el);
      if (el.dataset['data-resource'] === 'related-content') return;
      if (embed.resource === 'content-link') {
        return {
          kind: 'inline',
          type: 'link',
          data: embed,
          nodes: [
            {
              kind: 'text',
              text: embed['link-text']
                ? embed['link-text']
                : 'Ukjent link tekst',
            },
          ],
          isVoid: false,
        };
      }
      return {
        kind: 'block',
        type: 'embed',
        data: embed,
        isVoid: true,
      };
    },

    serialize(object) {
      if (!object.type || !object.type.startsWith('embed')) return;
      const data = object.data.toJS();
      const props = createEmbedProps(data);

      return <embed {...props} />;
    },
  },
];

export const topicArticeRules = topicArticeEmbedRule.concat(RULES);
export const learningResourceRules = RULES.concat(learningResourceEmbedRule);
