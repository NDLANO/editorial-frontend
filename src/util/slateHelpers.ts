/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Descendant, Node, Element, ElementType } from "slate";

export const findNodesByType = <T extends ElementType>(
  descendants: Descendant[],
  ...types: readonly T[]
): Extract<Element, { type: T }>[] => {
  return descendants
    .flatMap((descendant) => Array.from(Node.elements(descendant)))
    .map(([node]) => node)
    .filter((node): node is Extract<Element, { type: T }> => types.includes(node.type as T));
};
