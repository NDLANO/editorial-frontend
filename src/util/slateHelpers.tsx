/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Descendant, Node, Element } from 'slate';

export const findNodesByType = (descendants: Descendant[], ...type: Element['type'][]) => {
  const ret = descendants
    .flatMap((descendant) => Array.from(Node.elements(descendant)))
    .map(([node]) => node)
    .filter((node) => type.includes(node.type));
  return ret;
};
