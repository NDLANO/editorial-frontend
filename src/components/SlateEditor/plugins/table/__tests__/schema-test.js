/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Schema, State } from 'slate';
import tableSchema from '../schema';
import { tableSlateState } from '../../../../../util/__tests__/slateMockStates';
import { toJSON } from '../../../../../util/slateHelpers';

test('normalize table', () => {
  const state = State.fromJSON(tableSlateState);
  const schema = Schema.fromJSON(tableSchema);
  const change = state.change().normalize(schema);
  // const serializer = new Html({ rules: [tableRules], parseHtml: fragment });
  // const deserialized = serializer.deserialize(tableHTML);
  expect(toJSON(change.state)).toMatchSnapshot();
});
