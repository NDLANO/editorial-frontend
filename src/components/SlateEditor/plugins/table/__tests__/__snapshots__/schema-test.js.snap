// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize table 1`] = `
Object {
  "document": Object {
    "data": Object {},
    "kind": "document",
    "nodes": Array [
      Object {
        "data": Object {},
        "isVoid": false,
        "kind": "block",
        "nodes": Array [
          Object {
            "data": Object {},
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "data": Object {
                  "isHeader": true,
                },
                "isVoid": false,
                "kind": "block",
                "nodes": Array [
                  Object {
                    "kind": "text",
                    "ranges": Array [
                      Object {
                        "kind": "range",
                        "marks": Array [],
                        "text": "column 1",
                      },
                    ],
                  },
                ],
                "type": "table-cell",
              },
              Object {
                "data": Object {
                  "isHeader": true,
                },
                "isVoid": false,
                "kind": "block",
                "nodes": Array [
                  Object {
                    "kind": "text",
                    "ranges": Array [
                      Object {
                        "kind": "range",
                        "marks": Array [],
                        "text": "column 2",
                      },
                    ],
                  },
                ],
                "type": "table-cell",
              },
            ],
            "type": "table-row",
          },
          Object {
            "data": Object {},
            "isVoid": false,
            "kind": "block",
            "nodes": Array [
              Object {
                "data": Object {},
                "isVoid": false,
                "kind": "block",
                "nodes": Array [
                  Object {
                    "kind": "text",
                    "ranges": Array [
                      Object {
                        "kind": "range",
                        "marks": Array [],
                        "text": "column 1",
                      },
                    ],
                  },
                ],
                "type": "table-cell",
              },
              Object {
                "data": Object {},
                "isVoid": false,
                "kind": "block",
                "nodes": Array [
                  Object {
                    "kind": "text",
                    "ranges": Array [
                      Object {
                        "kind": "range",
                        "marks": Array [],
                        "text": "column 2",
                      },
                    ],
                  },
                ],
                "type": "table-cell",
              },
            ],
            "type": "table-row",
          },
        ],
        "type": "table",
      },
    ],
  },
  "kind": "state",
}
`;
